### DISCUSSION TITLE
Help with methods with CustomTypes

### DISCUSSION THREAD

**Discussion by siscia** [FreeOpcUa/python-opcua#1366] [Q&A]
Following up with #1364 

I am having a hard time creating a custom type and extracting its values.

For instance the following code:

```python
#!/usr/bin/env python3

import sys

from opcua import ua, Server
from opcua.common.type_dictionary_buider import DataTypeDictionaryBuilder

def square(parent, variant):
    print("square method call with parameters: ", variant.Value)
    variant.Value *= variant.Value
    return [variant]

def sum_of_squares(parent, variant):
    print("sum_of_square methos call with parameter: ", variant.Value)
    ## res = variant.Value.i**2 + variant.Value.j**2 # <- I was expecting this to work, somehow
    ret = 3*3 + 8*8
    return [ua.Variant(ret, ua.VariantType.Int64)]

if __name__ == "__main__":
    server = Server()
    server.create_custom_data_type(2, "ComplexType", description="A complex type")

    server.set_endpoint("opc.tcp://0.0.0.0:4840/")

    ns = server.register_namespace("http://gopcua.com/")
    dictBuilder = DataTypeDictionaryBuilder(server, ns, "http://gopcua.com/", "TestTypeBuilder")
    typeName = "ComplexType"
    dictBuilder.create_data_type(typeName)
    dictBuilder.add_field("i", ua.VariantType.Int64, typeName)
    dictBuilder.add_field("j", ua.VariantType.Int64, typeName)
    dictBuilder.set_dict_byte_string()
    server.load_type_definitions()
    
    main = server.nodes.objects.add_object(ua.NodeId("main", ns), "main")
    fnSquare = main.add_method(ua.NodeId("square", ns), "square", square, [ua.VariantType.Int64], [ua.VariantType.Int64])
    fnSumOfSquare = main.add_method(ua.NodeId("sumOfSquare", ns), "sumOfSquare", sum_of_squares, [ua.VariantType.ExtensionObject], [ua.VariantType.Int64])

    server.start()
```

I haven't found out yet how to reach the values in the `Complex` type.

I can see that the byte array that is received by the server is indeed of the right dimensions, (2 Int64) and that the type id is well parsed, but I am not understading how to read the values.

Any help is super appreciated.

**Comment by siscia** [comment]
Any help?

**Reply by AndreasHeine** [reply]
i made a example for those!

**Comment by oroulet** [comment]
What is you issue exactly? You receive an Extension Object instead of you custom struct?

Also use asyncua sync wrapper it is easier to use and has so far no known bugs for custom structures

**Reply by swamper123** [reply]
@siscia This repo is in maintainance mode and not state of the art. 
The latest code (with the mentioned sync wrapper) is in this repo: https://github.com/FreeOpcUa/opcua-asyncio

**Reply by siscia** [reply]
@oroulet  no, the problem is that I am not able to get the values inside the Extension Object.

So I (tried to) defined a complex structure, called `Complex` with two fields, `i` and `j`, both as Int64.
Now, in the method, I would like to access those fields.
So I would like to have a way to access those Int64.
Once I have access, I only need to square them, and sum them together.

@swamper123 Thanks! For the heads-up.
However, the code is used for testing another golang open source library, and I would prefer not to mess too much with the dependency of that library, especially since I believe it is just me missing how to read those values.

But you are right, the library should migrate its testing to the latest version of opcua-asyncio.

Hopefully I can allocate some time also for that migration.

**Reply by oroulet** [reply]
if load_data_type_definition() or load_type_defintion() is called then your mehtod should return an ua.ComplexType object and not an ExtensionObject object. If you get an ExtensionObject then something is wrong
